### PR TITLE
Handle ISO C <cmath> fixes and add example

### PR DESCRIPTION
**Pull request by ckormanyos**
This is the first attempt to correct ISO C signatures (first mentioned in #17) and specifically treated in #18

### CODE REVIEW

**Comment by ckormanyos** [comment]
The OP in #18 is still mentioning mismatches in signatures. So this still needs further investigation.
I'm rather sure the new example will stay unchanged. But i'd like to figure out what's going on with the OP's ATMEl Studio.

Cc: @salkinium and @chris-durand and @rleh

**Comment by ckormanyos** [comment]
Any feedback on this? Positive, negative, neutral?

**Comment by chris-durand** [comment]
> Any feedback on this? Positive, negative, neutral?

I'll take a look at it tomorrow. I am super busy right now and couldn't find the time yet.

**Comment by ckormanyos** [comment]
> take a look at it tomorrow

Thank you Chris (@chris-durand). This is not in a hurry.

There might, in fact, be a bit of discussion needed, as how best to proceed (or even if these changes are needed). I'm open for this potential discussion.

File: src/math.cc [hunk 1]
```diff
@@ -29,10 +33,12 @@ extern "C"
     return ::tan(x);
   }
 
+  #if defined(MODM_CMATH_GCC_VERSION) && (MODM_CMATH_GCC_VERSION < 110300L)
   float fabsf(float x)
   {
     return ::fabs(x);
   }
+  #endif
```
Comment (chris-durand) [inline src/math.cc:41]: I don't have an avr-gcc 11.3 at hand right now. What changes with regards to `fabsf` in the headers supplied by avr-gcc that we have to touch it? Is it a real function instead of a macro in 11.3?

File: src/math.cc [hunk 2]
```diff
@@ -129,17 +135,18 @@ extern "C"
     return ::pow(x, y);
   }
 
-  bool isnanf(float x)
+  int isnanf(float x)
```
[the remark this answers is not in the document]
Comment (ckormanyos) [reply src/math.cc:138]: > Changing the return types makes sense to me

I like that way of putting it. Indeed, I have struggled a lot in over the years on these kinds of points.

When wrapping the STL for `<cmath>` it seems that a certain amount of _empirical_ observation will always be necessary. This is because the compiler's `<math.h>` header file should be taken and left unchanged. The compiler supplier, however, will be advancing this C header and (hopefully) gettning closer and closer to standards adherence as we progress.

The bottom line is we need to keep an eye on the evolution of `<math.h>` and ensure that this is properly paired with `avr-libstdcpp`.

File: src/math.cc (hunk 1 above)
Comment (ckormanyos) [reply src/math.cc]: > What changes with regards to `fabsf` in the headers...

Good question. The three subroutines that I marked as _not_ _present_ in `math.cc` for the 11.3 compiler were (and need to be) marked that way since the designers of `<math.h>` elected to _inline` these functions at the C-header level by implementing them as `static` `inline` functions in the header itself. They can no longer be separate linkable entities in a C-file.

There is no way out of this one on the 11.3 port unless one would (as i do not recommend) modify `<math.h>` directly.

On the other hand, it would be conceptually thinkable to deliver a `math.h` with `avr-libstdcpp`, but I would not find that step necessary quite yet.

**Comment by chris-durand** [comment]
@ckormanyos We are running into the same issue of duplicate definitions of `fabs` and `isfinitef` with the updated Mac OS gcc 10.3 build. Do you know for which versions your fix needs to be applied? I have tried changing `MODM_CMATH_GCC_VERSION < 110300L` to `MODM_CMATH_GCC_VERSION < 100300L` and strangely enough, `fabsf` still compiles fine with 11.2.

**Comment by chris-durand** [comment]
> Do you have any way of posting part or all of the <math.h> header in question on the Mac OS versiion of avr-gcc that you are working on?

That is the file: https://gist.github.com/chris-durand/03d8c8555d33aeec17269e6ae25daf75
It does not actually define `fabsf`. Hm, strange...

**Comment by chris-durand** [comment]
That was not the right file, this one is: https://gist.github.com/chris-durand/7b6f51f6fcc527e46483471e69cae498

**Comment by chris-durand** [comment]
I have checked builds of avr-gcc 10.2, 10.3, 11.2 and 12.1. 10.2 and 11.2 have the macro definitions for `fabsf`. 10.3 and 12.1 have real functions. So we also have remove the definitions for gcc 10 with minor version >= 3.

**Comment by chris-durand** [comment]
The `math.h` definitions changed in avr-lib 2.1.0 and is independent of the gcc version. I have replaced the gcc version check with a check for `__AVR_LIBC_VERSION__` defined in `<avr/version.h>`.

**Comment by chris-durand** [comment]
@rleh @salkinium We have to merge this before the 2022q2 release.

**Review by chris-durand — APPROVED** [review]
Thanks @ckormanyos

**Review by salkinium — APPROVED** [review]
Lets do it

**Review by rleh — APPROVED** [review]

**Comment by ckormanyos** [comment]
Thanks folks @chris-durand, @rleh and @salkinium. For the moment, this seems to generally improve the situation regarding consistency of `<math.h>` and this STL-port's `<cmath>`.

If we need to refine any of these signatures, or if small inconsistencies that could be managed better happen to be found in the future, please let's not hesitate to correct them as we go.
